### ISSUE TITLE
Explicit commit of implicit transaction

### ISSUE THREAD

**Issue by mudrd8mz** [edwardspec/mediawiki-moderation#26]
MediaWiki: 1.27.2
Moderation: 1.0.45
PHP: 7.0
MySQL: 5.5.4

It seems that the `ModerationEditHooks::onPageContentSave()` causes a debugging notice being thrown:
```
( ! ) Notice: DatabaseBase::commit: Explicit commit of implicit transaction. Something may be out of sync! [Called from DatabaseBase::commit in .../mediawiki/includes/db/Database.php at line 2681] in .../mediawiki/includes/debug/MWDebug.php on line 300
Call Stack
#   Time    Memory  Function    Location
1   0.0000  377824  {main}( )   .../index.php:0
2   0.0249  2080056 MediaWiki->run( )   .../index.php:43
3   0.0249  2080056 MediaWiki->main( )  .../MediaWiki.php:509
4   0.0260  2265200 MediaWiki->performRequest( )    .../MediaWiki.php:735
5   0.0262  2275824 MediaWiki->performAction( ) .../MediaWiki.php:288
6   0.0262  2275920 SubmitAction->show( )   .../MediaWiki.php:493
7   0.0262  2275920 EditAction->show( ) .../SubmitAction.php:38
8   0.0263  2296320 EditPage->edit( )   .../EditAction.php:59
9   0.0846  2477064 EditPage->attemptSave( )    .../EditPage.php:578
10  0.0846  2477064 EditPage->internalAttemptSave( )    .../EditPage.php:1395
11  0.0856  2662096 WikiPage->doEditContent( )  .../EditPage.php:2043
12  0.0858  2663184 Hooks::run( )   .../WikiPage.php:1616
13  0.0859  2664280 ModerationEditHooks::onPageContentSave( )   .../Hooks.php:195
14  0.1570  3149696 DatabaseBase->commit( ) .../ModerationEditHooks.php:176
15  0.1571  3149824 wfWarn( )   .../Database.php:2681
16  0.1571  3149824 MWDebug::warning( ) .../GlobalFunctions.php:1141
17  0.1571  3150400 MWDebug::sendMessage( ) .../MWDebug.php:155
18  0.1572  3150656 trigger_error ( )   .../MWDebug.php:300
```

**Comment by edwardspec** [comment]
The notice itself is harmless (caused by correct code).
I'm investigating how to avoid it.

The commit() that caused this warning is intentional. It's a safeguard against the following scenario:
1) some third-party extension calls doEditContent(),
2) this third-party extension assumes that "moderation-edit-queued" is an error,
3) it throws MWException, causing current transaction to be rolled back,
4) rollback removes the newly added row in "moderation" table.

**Comment by mudrd8mz** [comment]
Thanks for paying attention to this so promptly Edward! I had similar suspicion. The remaining issue for us is that this harmless notice outputs early before other HTTP headers are sent (with all the consequences) and it makes our staging site (with debugging enabled) unusable by testers.

**Comment by edwardspec** [comment]
As a quick workaround, you can replace the following line in `hooks/ModerationEditHooks.php`:
`$dbw->commit();`
with
`$dbw->commit( __METHOD__, 'flush' );`

That will suppress the notice.

**Comment by mudrd8mz** [comment]
Many thanks, will give it a try.

**Comment by mudrd8mz** [comment]
Our QA team reported the suggested change fixes the problem with warning for us.

**Comment by edwardspec** [comment]
Fixed in 1.0.46.

I removed this commit() call completely.
Instead we detect a database rollback and (if it happens) save the queued edit again.

**Comment by mudrd8mz** [comment]
Many thanks.
